### PR TITLE
[SLP]Improve minbitwidth analysis for operands of IToFP and ICmp instructions.

### PR DESCRIPTION
**Pull request by alexey-bataev**
Compiler can improve analysis for operands of UIToFP/SIToFP instructions
and operands of ICmp instruction.

### CODE REVIEW

**Comment by alexey-bataev** [comment]
Ping!

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -3662,7 +3662,7 @@ class BoUpSLP {
   std::optional<std::pair<unsigned, unsigned>> CastMaxMinBWSizes;
 
   /// Indices of the vectorized trunc nodes.
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:3664]: Update comment?

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp]: Done!

**Comment by alexey-bataev** [comment]
Ping!

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 2]
```diff
@@ -6588,7 +6589,13 @@ void BoUpSLP::buildTree_rec(ArrayRef<Value *> VL, unsigned Depth,
                 PrevMaxBW),
             std::min<unsigned>(DL->getTypeSizeInBits(VL0->getType()),
                                PrevMinBW));
-        TruncNodes.insert(VectorizableTree.size());
+        ExtraBitWidthNodes.insert(VectorizableTree.size() + 1);
+      } else if (ShuffleOrOp == Instruction::SIToFP ||
+                 ShuffleOrOp == Instruction::UIToFP) {
+        if (ComputeNumSignBits(VL0->getOperand(0), *DL, 0, AC, nullptr, DT) *
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:6595]: Pull the ComputeNumSignBits calls out so the if line aren't distorted so much by clang-format?

**Review by RKSimon — APPROVED** [review]
LGTM with one minor

**Comment by mikaelholmen** [comment]
Hi @alexey-bataev 

I've bisected a miscompile back to this patch.
Reproduce with
```
opt -passes=slp-vectorizer bbi-94784.ll -S -o - -mtriple=aarch64 -slp-threshold=-10 -slp-vectorize-hor=0
```
[bbi-94784.ll.gz](https://github.com/llvm/llvm-project/files/15064102/bbi-94784.ll.gz)

For input %in1=0x3 and %in2=ffff the input function returns 0, but after slp-vectorizer with this patch is returns 2.
I think the problem is that after vectorization the "mul" operates on i16 values but it should really be on i64 otherwise the compares with 196605 will go wrong.

**Comment by alexey-bataev** [comment]
> Hi @alexey-bataev
> 
> I've bisected a miscompile back to this patch. Reproduce with
> 
> ```
> opt -passes=slp-vectorizer bbi-94784.ll -S -o - -mtriple=aarch64 -slp-threshold=-10 -slp-vectorize-hor=0
> ```
> 
> [bbi-94784.ll.gz](https://github.com/llvm/llvm-project/files/15064102/bbi-94784.ll.gz)
> 
> For input %in1=0x3 and %in2=ffff the input function returns 0, but after slp-vectorizer with this patch is returns 2. I think the problem is that after vectorization the "mul" operates on i16 values but it should really be on i64 otherwise the compares with 196605 will go wrong.

Thanks for the reproducer, will double check

**Comment by alexey-bataev** [comment]
> ```
> -mtriple=aarch64 -slp-threshold=-10 -slp-vectorize-hor=0
> ```

Must be fixed in 102a8110943bf18fab923e01610a693587a01945

**Comment by mikaelholmen** [comment]
> > ```
> > -mtriple=aarch64 -slp-threshold=-10 -slp-vectorize-hor=0
> > ```
> 
> Must be fixed in [102a811](https://github.com/llvm/llvm-project/commit/102a8110943bf18fab923e01610a693587a01945)

Yep, thanks

**Pull request merged**